File: 3351-maximize-happiness-of-selected-children/maximize-happiness-of-selected-children.cpp
class Solution {
public:
    long long maximumHappinessSum(vector<int>& happiness, int k) {
        sort(happiness.begin(), happiness.end(), std::greater<int>());
        long long total = 0; 
        for (int i = 0; i < k; i++){
            total += max(0, happiness[i] - i); 
        }
        return total; 
    }
};

File: 3351-maximize-happiness-of-selected-children/maximize-happiness-of-selected-children.py
"""
0:00 - understanding
choose k children in k turns, where we pick one child per turn 

each pick, all non picked children happiness decreases by 1 until 0 

need maximum sum of selecting k children 

[1, 2, 3]
k = 2 

it would be wisest to greedily select the largest children

"""




class Solution:
    def maximumHappinessSum(self, happiness: List[int], k: int) -> int:
        return sum(max(0, score - i) for i, score in enumerate(sorted(happiness, reverse=True)[0:k]))